### ISSUE TITLE
Messages aren't published until connection.close() is called.

### ISSUE THREAD

**Issue by ianchadwick** [CopernicaMarketingSoftware/AMQP-CPP#184]
Hi there, I'm having some problems with publishing messages using libevent. The code I'm using is pretty close to the code in the example.

What I want to happen and would expect is to be able to publish new messages at will and have each call to publish actually send the message to the exchange. The code below doesn't do this, instead everything is queued and then pushed when connection->close() is called.

How do I make it push each item individually? Thanks!

```c++
void onConnected(AMQP::TcpConnection *connection) override 
    {
        AMQP::TcpChannel channel(connection);

        for (int i = 0; i < 10; i++) {
            // I would expect these to be published as this is called, but they are now.
            channel.publish("test", "", "hello-there");
        }

        // the queue is flushed now... =/
        connection->close();
    }
```

**Comment by EmielBruijntjes** [comment]
Where do you run the event loop?

**Comment by ianchadwick** [comment]
It's run after setting up the handler:

```c++
auto eventBase = event_base_new();
QueueHandler handler = QueueHandler(eventBase);
AMQP::Address address(queueConfig);
AMQP::TcpConnection connection(&handler, address);
AMQP::TcpChannel channel(&connection);

... some channel onError & onReady stuff..

event_base_dispatch(eventBase);
```

**Comment by EmielBruijntjes** [comment]
You should not close the connection. Just let the event loop run, the library will send the messages to RabbitMQ when the file descriptor becomes writable.

**Comment by ianchadwick** [comment]
Unfortunately it's not sending them hence the contrived example with the connection->close().

It almost feels as though event loop is being blocked by my loop. Is there anything I could be doing which would do that?

Here's a version using just the default libevent handler which only sends the messages once the loop is exits. Why doesn't the loop allow the messages to be sent?

Thanks for your help with this! :D

```c++
    auto eventBase = event_base_new();
    AMQP::LibEventHandler handler(eventBase);
    AMQP::Address address("amqp://...");
    AMQP::TcpConnection connection(&handler, address);
    AMQP::TcpChannel channel(&connection);

    channel.onReady([&channel, &connection]() {
        for(int i = 0; i < 10; i++) {
            // publish and sleep to highlight the problem
            channel.publish("test", "", "hello-there");
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
        }
    });

    event_base_dispatch(eventBase);
```

**Comment by EmielBruijntjes** [comment]
I do not know much about libevent. But with libev, the following code simply works:

````
/**
 *  DeclareExchange.cpp
 *
 *  Example program to declare an exchange
 */
 
/**
 *  Dependencies
 */
#include <ev.h>
#include <amqpcpp.h>
#include <amqpcpp/libev.h>

/**
 *  Main procedure
 *  @return int
 */
int main() 
{
    // create event loop
    auto *loop = EV_DEFAULT;
    
    // construct handler
    AMQP::LibEvHandler handler(loop);
    
    // create tcp connection
    AMQP::TcpConnection connection(&handler, AMQP::Address("amqp://localhost/"));
    
    // create a channel
    AMQP::TcpChannel channel(&connection);
    
    // wait until the channel is ready
    channel.onReady([&channel]() {
        
        // declare an exchange
        channel.declareExchange("myexchange");
        
        // create aqueue
        channel.declareQueue("myqueue");
        
        // bind exchange to queue
        channel.bindQueue("myexchange", "myqueue", "routingkey");

        // loop to publish multiple messages
        for(int i = 0; i < 10; i++) 
        {
            // publish and sleep to highlight the problem
            channel.publish("myexchange", "routingkey", "hello-there");
        }        
    });
    
    // run the event loop
    ev_run(loop, 0);
    
    // done
    return 0;
}

````

**Comment by ianchadwick** [comment]
I've just given that a go and it's still doing the same thing.

I copy and pasted the above code just adding a sleep after each publish so it's easier to see them coming through individually (or not as the case is).

With this sleep a message should be being sent every 250ms but as you can see it clearly waits 2.5s before it sends all messages at once (From No. 21 onwards).

![screen shot 2018-02-06 at 21 22 22](https://user-images.githubusercontent.com/733223/35884880-f396ca82-0b83-11e8-8a09-b9c82e36773b.png)

**Comment by ianchadwick** [comment]
Fairly sure the for loop is blocking the event loop from continuing. If I add the loop in to a new thread (which isn't safe) it sends each publish immediately.

This kind works, except it's not thread safe (this is an assumption based on it SIGABRTing sporadically when going flat out).

```c++
// ... setup the handler, connection and channel ...

std::thread t([&channel, &loop]() {
    channel.publish("myexchange", "", "hello-there");
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
});

ev_run(loop, 0);
```

As my understanding of libevnt & libev increases it seems this usage is not going to be possible as anything that needs to loop will simply block the event loop until it finishes. Is this your understanding too? Have you ever needed to do what I'm trying to do by having messages published immediately?

Thanks again for your help with this.

**Comment by EmielBruijntjes** [comment]
You have a case. RabbitMQ has a little odd behavior: if you send it multiple instructions in a row, it behaves strange when it already receives the next instruction while it is still busy processing the previous instruction. I don't know why this is, maybe it handles each instruction in a separate thread or so, and it cannot cope if one thread is busy declaring a queue while the next one is already handling a message to publish to this thread.

On the other hand, AMQP-CPP has a very tolerant API: it allows you to send all the instructions in row, without worrying about previous steps or handshakes that still have to be finished. You can see that in the example that I sent: an exchange and queue were declared, bound and messages were published to it, without any wait time between these calls. To make this possible, AMQP-CPP has an internal pipeline in which it stores future instructions that cannot yet be sent to RabbitMQ. From your program standpoint AMQP-CPP acts like all instructions are sent right away, but under the hood the messages are cached until previous answers are received and the event loop reports that the socket is writable again. For you program this means that sometimes when you call the channel.publish() method the data is sent to RabbitMQ right away, and sometimes it is first cached and you have to run the event loop for a while before the instruction gets sent to RabbitMQ.

However, I see that AMQP-CPP is a little too conservative here. In your case it was caching outgoing messages even though it was well possible to send them right away. That was because the channel "onReady" callback was called while the "use cache instead of sending data right away" boolean was not yet reset. I'll fix this.

**Comment by dichen16** [comment]
guys, how can i resue the connection after I execute the "ev_run"?
if I use a thread only for publishing mesasage, after building the channel and run the "ev_run", 
can I use the connection to publish message

**Comment by EmielBruijntjes** [comment]
Not enough information / question is too vague

**Comment by dichen16** [comment]
`void RabbitPublishThread::Connect()
{
    m_pTcpConnection = std::unique_ptr<AMQP::TcpConnection>(
        new AMQP::TcpConnection(this, AMQP::Address("amqp://hzhs:hzhs@192.168.0.140:5672")));
    m_pTcpChannel = std::unique_ptr<AMQP::TcpChannel>(new AMQP::TcpChannel(m_pTcpConnection.get()));
    ev_run(m_loop, 0);
}

void RabbitPublishThread::run()
{
    m_thread = std::thread(std::bind(&RabbitPublishThread::Connect, this));
    m_thread.detach();
    // the following publish did not work
    m_pTcpChannel->publish("my-exchange", "my-key", "my-payload");
}`
Hi, Emiel ,my question is after detach the thread,  my loop does not work, i am confused.

**Comment by EmielBruijntjes** [comment]
I think you need a lot to learn about C++ fundamentals first. Your code makes no sense at many levels.

**Comment by dichen16** [comment]
Could you tell me which part result to the unwilling situation?
You are right, I do need to learn more about C++ fundamentals.

**Comment by dichen16** [comment]
I just found which part result in the unwilling situation, my fault
